GAWB-992: use GCS signed urls for downloading

davidangb opened the pull request:
Related PRs are already merged:
broadinstitute/firecloud-develop#138
broadinstitute/firecloud-develop#139

Previously, when downloading a file from GCS, we proxied the entire file through PerRequest in orchestration. For large files, this was hugely inefficient, and according to the jira ticket could crash the orchestration service.

Now, when the user requests a download, we 1) request the user's identity from Google, using their token; 2) perform a final check to see if the user has permission access to the file. If the user is valid according to Google, and has permission to the file, we immediately sign a URL and redirect the browser to that signed URL. The URL is time-limited to two\* minutes for safety - and since we immediately redirect to it, two minutes should be more than enough for the current user. Empirical testing shows that the time limit is only relevant for _starting_ the download; once the download is going, it won't stop midway because of the signing expiration.

We log (in the orchestration log, which gets pushed to loggly) all download attempts, both successes and failures.

This has FISMA/security impact. If the current user shares the pre-signed url with somebody else, that other person is able to download the file within the two-minute time period, even if they would not otherwise have permission to that file.

This PR also fixes GAWB-988.

_* original PR had a five-minute expiration; I changed this to two minutes in the fourth commit._
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  - Acceptance criteria exists and is met
  - Note any changes to implementation from the description
  - Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

davidangb committed:
Use signed URLs for GCS downloads. Requires new service account in orch.

davidangb committed:
use rawls service account for signing urls

davidangb committed:
check the user's access to the GCS file before signing

dmohs commented on src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -123,4 +125,39 @@ object HttpGoogleServicesDAO {
+    val verb = "GET"
+    val md5 = ""
+    val contentType = ""
+    val expire = (System.currentTimeMillis() /1000) + 300 // expires 5 minutes (300 seconds) from now
```
Nit: spacing. Also, I find it helpful to include the unit for integer durations, so I'd call this `expireSeconds`. Your option.

dmohs commented on src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -123,4 +125,39 @@ object HttpGoogleServicesDAO {
+    signature.update(signableString.getBytes("UTF-8"))
+    val signedBytes = signature.sign()
+
+    // assemble the final url
```
Nit: obvious. Suggest removing comment.

dmohs commented:
👍  Security-wise, I see no risks beyond what has already been stated.

However, we can use the functionality introduced here to improve our security posture in general. As a follow-up to this work, we can provide a separate API call for getting a signed URL that uses the `Authorization` header instead of a cookie. Once we update the UI to use that endpoint, we can likely stop storing the token in a cookie altogether.

davidangb committed:
decrease signed url validity to 2 minutes; minor syntax/var name changes

davidangb commented:
updated spacing and variable naming. I left the obvious comment in place, because it's consistent with the stucture of that method - every other block is commented.

I also changed the expiration time to two minutes!

I don't think we can have a follow-on story to use the Authorization: header - at least not without significant work. The big problem here is that the download (or for files like PDFs, the opening) of the target file is in a `target="_blank"` new window. We don't have a way, due to browser limitations, to tell that new window to send an Authorization: header.

davidangb committed:
check user's info in Google before handing out signed URL; log all downloads (success or fail)

davidangb committed:
make hardcoded url final

dvoet commented:
regarding the 2 minute expiration and starting the download - as downloads become larger it becomes more likely that they will fail which may lead us to multi-part downloads and then 2 minutes will be a problem... but that may be for another day. However, since the whole point of this is to download large files, maybe it is today.

dvoet commented on src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -123,4 +127,39 @@ object HttpGoogleServicesDAO {
+    val signature = java.security.Signature.getInstance("SHA256withRSA")
+    signature.initSign(privateKey)
+    signature.update(signableString.getBytes("UTF-8"))
+    val signedBytes = signature.sign()
```
Is there no way to sign as the user? The obvious problem is security which you have jumped through some hoops deal with. The other problem is that this won't work if the rawls service account does not have access while the user does (protected TCGA data may fall into this category, I am not sure).

dmohs commented on src/main/scala/org/broadinstitute/dsde/firecloud/service/CookieAuthedService.scala:
```diff
@@ -37,16 +40,41 @@ trait CookieAuthedService extends HttpService with PerRequestCreator with FireCl
+
+            val objectStr = s"gs://$bucket/$obj"
+
+            userPipeline{userReq} map { userResponse =>
```
Is the reason for checking for a valid user so we can log invalid access requests? If so, it would be helpful to see that in a comment too.

dmohs commented:
One comment suggestion. No need to re-review.

We could fetch the GCS URL and make that the `href` attribute of the link, so it would work in the target="_blank" case.

Re Doug's comment: Can we move forward with this and see how it goes? Even if it prevents downloading of certain files, not having the server crash is pretty important.

davidangb committed:
WIP: also check service account's access to the file

davidangb committed:
refactor, move logic into HttpGoogleServicesDAO

davidangb commented:
@dmohs, @dvoet : I've added functionality to check if the service account has permission to sign the url, and I've refactored. The second-to-last commit adds the functionality, and the last commit should be pure refactoring.

In the case where the user has permission, but the service account doesn't, we revert to using a direct link - that's the link that could potentially use the wrong identity.

David Mohs - we can't fetch the link and use it in the `href=`. In that model, we have no control over the delay between generating the link and using the link - which means our signed urls would be prone to expire.

dvoet commented:
👍

dmohs commented:
👍  Very reasonable handling of a difficult situation.
